feat: Additional Features (#1345)
discordjs/guide: guide/popular-topics/reloading-commands.md

This section assumes you did follow the [Command Handling](/guide/creating-your-bot/command-handling.md) part.

```js {1}
const { SlashCommandBuilder } = require('discord.js');

module.exports = {
	data: new SlashCommandBuilder()
		.setName('reload')
		.setDescription('Reloads a command.')
		.addStringOption(option =>
			option.setName('command-name')
			.setDescription('Insert the command name.')
			.setRequired(true)),
	async execute(interaction) {
		// ...
	},
};
```

In this command, you will be using a command name. First off, you need to check if the command you want to reload exists. You can do this check similarly to getting a command in your main file:

```js {4-6,8-10}
module.exports = {
	// ...
	async execute(interaction) {
		const commandName = interaction.options.getString('command-name', true).toLowerCase();
		const command = interaction.client.commands.get(commandName);

		if (!command) {
			return interaction.reply(`There is no command with name \`${commandName}\`!`);
		}
	},
};
```

> To build the correct file path, you will need the file name. You can use `command.data.name` for doing that.

Reviewer: ```suggestion
To build the correct file path, you will need the file name. You can use `command.data.name` for doing that.
```
this is incorrect if the user doesn't has file names same as the command names

Author: That's what the line 5 is for.

Reviewer: the guide still recommends a /commands folder though so might be better to have all paths going there

# Reloading commands

When writing your commands, you may find it tedious to restart your bot every time you want to test even your code's slightest change. However, a single bot command can reload commands if you have a command handler.

This section assumes you did follow the [Command Handling](/guide/creating-your-bot/command-handling.md) part.

```js {1}
const { SlashCommandBuilder } = require('discord.js');
const fs = require('node:fs');

module.exports = {
	data: new SlashCommandBuilder()
        .setName('reload')
        .setDescription('Reloads a command.')
        .addStringOption(option =>
            option.setName('command-name')
            .setDescription('Insert the command name.')

>             .setRequired('true')),

Reviewer 2: ```suggestion
            .setRequired(true)),
```

```js {4-6,8-10}
module.exports = {
	// ...
	async execute(interaction) {
		const commandName = interaction.options.getString('command-name', true).toLowerCase();
		const command = interaction.client.commands.get(commandName);

		if (!command) {
			return interaction.reply(`There is no command with name \`${commandName}\`!`);
		}
	},
};
```

To build the correct file path, you will need the file name. You can use `command.data.name` for doing that.

In theory, all there is to do is delete the previous command from `client.commands` and require the file again. In practice, you cannot do this easily as `require()` caches the file. If you were to require it again, you would load the previously cached file without any changes. You first need to delete the file from the `require.cache`, and only then should you require and set the command file to `client.commands`:

```js {3,5-12}
delete require.cache[require.resolve(`./${command.data.name}.js`)];

try {
    interaction.client.commands.delete(command.data.name);

> 	const newCommand = require(`.${command.data.name}.js`);

Reviewer 2: This is wrong

# Reloading Commands

> When writing your commands, you may find it tedious to restart your bot every time for testing the smallest changes, With a command handler, you can eliminate this issue and reload your commands while your bot is running.

Reviewer 3: ```suggestion
When writing your commands, you may find it tedious to restart your bot every time for testing the smallest changes. With a command handler, you can eliminate this issue and reload your commands while your bot is running.
```

> This section assumes you did follow the [Command Handling](/guide/creating-your-bot/command-handling.md) part.

Reviewer 3: ```suggestion
This section assumes you followed the [Command Handling](/guide/creating-your-bot/command-handling.md) part.
```

Reviewer: "the Command Handling part" doesn't seems exactly right, what are the thoughts on "the Command Handling section", or maybe the standard TIP could be better here "This page is a follow-up to the Command Handling page. Please carefully read it first so that you can understand the methods used in this section."

Reviewer 3: "section" sounds good as does the generic tip.

> ::: warning
> The require method does not work in ESM. If you are using it, try out the [hot-esm](https://www.npmjs.com/package/hot-esm) module.
> :::

Reviewer: ```suggestion
::: warning
ESM does not support require and clearing import cache, you can use [hot-esm](https://www.npmjs.com/package/hot-esm) to import files without cache.
:::
```
esm users would already know about require not working so it's better to give a heads up that import cache is not manageable either

> ESM does not support require and clearing import cache, you can use [hot-esm](https://www.npmjs.com/package/hot-esm) to import files without cache.

Reviewer 3: ```suggestion
ESM does not support require and clearing import cache. You can use [hot-esm](https://www.npmjs.com/package/hot-esm) to import files without cache.
```
